ocaml/lib/snarky_js_bindings_lib.ml

```diff
@@ -1248,563 +1219,163 @@ let poseidon =
-      < snarkyMain : ('w -> 'p -> unit) Js.callback Js.prop
-      ; snarkyWitnessTyp : 'w as_field_elements Js.t Js.prop
-      ; snarkyPublicTyp : 'p as_field_elements Js.t Js.prop >
-      Js.t
-  end
+  let as_prover = Impl.as_prover
 
-  let main_and_input (type w p) (c : (w, p) Circuit_main.t) =
-    let main ?(w : w option) (public : p) () =
-      let w : w =
-        witness c##.snarkyWitnessTyp
-          (Js.wrap_callback (fun () -> Option.value_exn w))
-      in
-      Js.Unsafe.(fun_call c##.snarkyMain [| inject w; inject public |])
-    in
-    (main, typ_ c##.snarkyPublicTyp)
+  let run_and_check (f : unit -> unit) =
+    try
+      Impl.run_and_check_exn (fun () ->
+          f () ;
+          fun () -> () )
+    with exn -> raise_exn exn
 
-  let generate_keypair (type w p) (c : (w, p) Circuit_main.t) :
-      keypair_class Js.t =
-    let main, input_typ = main_and_input c in
-    let cs =
-      Impl.constraint_system ~input_typ ~return_typ:Snark_params.Tick.Typ.unit
-        (fun x -> main x)
-    in
-    let kp = Impl.Keypair.generate ~prev_challenges:0 cs in
-    new%js keypair_constr kp
+  let run_unchecked (f : unit -> unit) =
+    try
+      Impl.run_and_check_exn (fun () ->
+          Snarky_backendless.Snark0.set_eval_constraints false ;
+          f () ;
+          Snarky_backendless.Snark0.set_eval_constraints true ;
+          fun () -> () )
+    with exn -> raise_exn exn
 
   let constraint_system (main : unit -> unit) =
     let cs =
-      Impl.constraint_system ~input_typ:Impl.Typ.unit
-        ~return_typ:Snark_params.Tick.Typ.unit (fun () -> main)
-    in
-    let rows =
-      Kimchi_pasta_constraint_system.Vesta_constraint_system.get_rows_len cs
-    in
-    let digest =
-      Backend.R1CS_constraint_system.digest cs |> Md5.to_hex |> Js.string
-    in
-    let json =
-      Js.Unsafe.(
-        fun_call
-          global ##. JSON##.parse
-          [| inject (Backend.R1CS_constraint_system.to_json cs |> Js.string) |])
+      Impl.constraint_system ~input_typ:Impl.Typ.unit ~return_typ:Impl.Typ.unit
+        (fun () -> main)
     in
     object%js
-      val rows = rows
+      val rows = Backend.R1CS_constraint_system.get_rows_len cs
+
+      val digest =
+        Backend.R1CS_constraint_system.digest cs |> Md5.to_hex |> Js.string
 
-      val digest = digest
+      val json =
+        Backend.R1CS_constraint_system.to_json cs |> Js.string |> json_parse
+    end
 
-      val json = json
+  module Circuit = struct
+    module Main = struct
+      let of_js (main : public_input_js -> unit) =
+        let main' public_input () = main (Public_input.to_js public_input) in
+        main'
     end
 
-  let prove (type w p) (c : (w, p) Circuit_main.t) (priv : w) (pub : p) kp :
-      proof_class Js.t =
-    let main, input_typ = main_and_input c in
-    let pk = Keypair.pk kp in
-    let p =
-      Impl.generate_witness_conv ~return_typ:Snark_params.Tick.Typ.unit
+    let compile main public_input_size =
```

Comment:
when we talk about compile here, what does this mean? Is this Kimchi compiling or translating the circuit into a constraint system?

Reply:
yes, both steps are here:
* `Impl.constraint_system` is running the function and recording constraints & variables
* `Impl.Keypair.generate` is Kimchi creating the prover and verification keys from the constraint system